packages/core/utils/lib/async.js

```diff
@@ -0,0 +1,55 @@
+'use strict';
+
+const { chunk } = require('lodash/fp');
+
+function pipeAsync(...methods) {
+  return async (data) => {
+    let res = data;
+
+    for (const method of methods) {
+      res = await method(res);
+    }
+
+    return res;
+  };
+}
+
+/**
+ * @type { import('./async').mapAsync }
+ */
+function mapAsync(promiseArray, { concurrency = Infinity } = {}) {
+  const appliedConcurrency = concurrency > promiseArray.length ? promiseArray.length : concurrency;
+  const promiseArrayChunks = chunk(appliedConcurrency)(promiseArray);
+
+  return async (iteratee) => {
+    return promiseArrayChunks.reduce(async (prevChunksPromise, chunk, chunkIndex) => {
+      // Need to await previous promise in order to respect the concurrency option
+      const prevChunks = await prevChunksPromise;
+      // As chunks can contain promises, we need to await the chunk
+      const awaitedChunk = await Promise.all(chunk);
+      const transformedPromiseChunk = await Promise.all(
+        // Calculating the index based on the original array, we do not want to have the index of the element inside the chunk
+        awaitedChunk.map((value, index) => iteratee(value, chunkIndex * appliedConcurrency + index))
+      );
+
+      return prevChunks.concat(transformedPromiseChunk);
+    }, Promise.resolve([]));
+  };
+}
```

Comment:
I think doing it by chunk is not optimal as the optimal way would be to start a new promise right after one was resolved. There is a little lib called p-map that we already use that can be useful for that. Something like:
```js
const { curry } = require('lodash/fp');
const pmap = require('p-map');

const mapAsync = curry((iteratee, options, array) => {
  return pmap(array, iteratee, options);
});
```

WDYT?

Reply:
The lib is this one : https://github.com/sindresorhus/p-map
We use it in `packages/core/upload/server/controllers/admin-file.js` for instance. only version 4.0.0 for the moment as the versions above 5 removed the support for `require`.
This lib also has a system to handle errors properly

Reply:
When the concurrency is Infinity, the user can just do a classic `Promise.all`, so I guess `mapAsync` would be used instead when there is a defined concurrency. I would put 1 by default as it's our initial usecase and as when there is a concurrency != 1, then only the user knows what concurrency they want.

Reply:
> When the concurrency is Infinity, the user can just do a classic Promise.all, so I guess mapAsync would be used instead when there is a defined concurrency. I would put 1 by default as it's our initial usecase and as when there is a concurrency != 1, then only the user knows what concurrency they want.

I think that we want to keep the Infinity default base, as it is still a map around an async array / iteratee func. For sure you can use `Promise.all();` to resolve an array of promises but using the `.map()` function (`Array.map`) won't handle the values in the array the same way as mapAsync (mapAsync is awaiting values inside the array before passing it to the iteratee). Also, I think that if you don't send explicit concurrency option is saying that you don't want any concurrency (I have based my thinking on the Promise.map function of Bluebird lib).

Didn't know about p-map, will dig a bit more in this lib to know if it does everything that is wanted on this function.


I don't understand why you say `I think doing it by chunk is not optimal as the optimal way would be to start a new promise right after one was resolved.`. Isn't the code doing it ? Waiting for previous chunk to be processed and then doing the next one ?

Reply:
I understand the reasoning for the default value, I don't have a strong opinion. p-map is using a default value of `100`. Which I also like because it's like Infinity but with a safeguard.

From what I understand of the code, if the concurrency is 50, it will wait for all the 50 promises of the chunk to be resolved before starting the next chunk of 50, right?
The number of promises running is then not always at 50. Whereas if you launch 50 promises and then launch a new promise right after one is resolved, then you always have 50 promises running (except at the end of course) which is faster.

<img width="350" alt="Chunk by chunk" src="https://user-images.githubusercontent.com/11708220/209092974-57f18ef1-6111-4b88-9cb2-8ac639890040.png"> <img width="330" alt="Promise runs just after one is over with a max of 50" src="https://user-images.githubusercontent.com/11708220/209092986-82563627-1dd0-46d6-84b4-079e0541540a.png">

Reply:
nice 😊